src/Proto3/Wire/Decode.hs

```diff
@@ -127,7 +127,29 @@ data ParsedField = VarintField Word64
 -- fromList [(1,[6,3]),(2,[4])]
 --
 toMap :: [(FieldNumber, v)] -> M.IntMap [v]
-toMap kvs0 = M.fromListWith (<>) . map (fmap (:[])) . map (first (fromIntegral . getFieldNumber)) $ kvs0
+toMap kvs0 = makeMap . map (first (fromIntegral . getFieldNumber)) $ kvs0
+  where
+    makeMap :: [(Int, v)] -> M.IntMap [v]
+    makeMap = close . foldl' combineSeen Nothing
+
+    close Nothing = M.empty
+    close (Just (m, k, v)) = M.insertWith (++) k v m
+
+    -- If keys are in order, then we don't have to make any lookups,
+    -- we just maintain the active element.
+    -- Out of order keys will lookup in the map
+    combineSeen :: Maybe (M.IntMap [v], Int, [v]) -> (Int, v) -> Maybe (M.IntMap [v], Int, [v])
+    combineSeen Nothing (k1, a1) = Just (M.empty, k1, [a1])
+    combineSeen (Just (m, k2, as)) (k1, a1) =
+      if k1 == k2
+        then Just (m, k1, a1 : as)
+        -- It might seem that we want to use DList but we don't because:
+        -- - alter has worse performance than insertWith, and there's no upsert
+        -- - We're building up a list of elements in a recursive way
+        --    that will be opaque to GHC
+        -- - DList would add another dependency
```

Comment:
Also, there's no need for `DList` because `insertWith (++)` prepends the singleton list `[a1]`, so it's still O(1)

Reply:
Well the point is that it's not always a singleton

Reply:
Oh, whoops.  I misread the code.  Never mind